[Epoch Sync] EpochSyncProof validation

staffik opened the pull request:
Issue: https://github.com/near/nearcore/issues/11932

Implements EpochSyncProof validation, according to:
- [design doc](https://docs.google.com/document/d/14Itc9Hs7ewTRmcGANid9UCaYcJzaLzzM7FsvYYIFKDY/edit#heading=h.p2oiicluwd6o)
- [`epoch_sync.rs`](https://github.com/near/nearcore/blob/d6d6e4782f7d57a7b7acc2894e1668355012ff74/core/primitives/src/epoch_sync.rs)

Summary:
- Move block merkle proof retrieval logic from `Chain` to a separate trait (`MerkleProofAccess`). Make `Store` implements this trait so that we can `get_block_proof` while serving `EpochSyncRequest` from a separate thread and having no access to `Chain`.
- Refactor `verify_approval` and `compute_bp_hash` so that we can use it for proof validation.
- Derive `first_block_info_in_epoch` from block header.
- Implement the validation logic.
- Add 2 more fields to the proof
  - `EpochSyncProofPastEpochData::protocol_version`
  -  `merkle_proof_for_first_block` generated at height of final block of current epoch.
  - Apparently, we need to add even more to enable generating `merkle_proof_for_first_block` at the bootstrapped node. That will be done as a follow-up issue: https://github.com/near/nearcore/issues/12255.
- Create a bunch of follow-up issues (added to the [tracking issue](https://github.com/near/near-one-project-tracking/issues/73)):
  - https://github.com/near/nearcore/issues/12255
  - https://github.com/near/nearcore/issues/12256
  - https://github.com/near/nearcore/issues/12258
  - https://github.com/near/nearcore/issues/12259
  - https://github.com/near/nearcore/issues/12260

staffik commented on chain/client/src/sync/epoch.rs:
```diff
@@ -483,15 +472,20 @@ impl EpochSync {
         store_update.set_ser(
             DBCol::BlockOrdinal,
-            &borsh::to_vec(&proof.current_epoch.merkle_proof_for_first_block.size()).unwrap(),
+            &index_to_bytes(last_header.block_ordinal()),
```
Concern: `BlockHeader::block_ordinal()` is supported since `BlockHeaderV3`.

robin-near replied:
I don't think that's an issue. EpochSync to a block header version below V3 is unsupportable due to also not having epoch_data_hash.

robin-near replied:
But it will be worth making a test to check that trying to sync to a version below V3 will fail. I'll file an issue.

staffik commented on chain/client/src/sync/epoch.rs:
```diff
@@ -483,15 +472,20 @@ impl EpochSync {
+        let last_epoch_data = &proof.past_epochs[proof.past_epochs.len() - 2];
+        let last_epoch_last_final_block_height = last_epoch_data.last_final_block_header.height();
+        let first_block_info_in_epoch =
+            BlockInfo::from_header(&last_header, last_epoch_last_final_block_height);
```
Concern: there is also `from_header_and_endorsements()` method, but `from_header` would copy endorsements from the header. cc @tayfunelmas

staffik commented:
Looking into failing integration tests.

robin-near commented on chain/chain/src/store/merkle_proof.rs:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    /// Reconstruct node at given position (index, level). If the node does not exist, return `None`.
+    fn reconstruct_merkle_tree_node(
```
Consider moving this and some of the above functions outside of the Trait. Otherwise, these functions become public and pollute the namespace of what's available on Store. You can pass in the &impl trait in place of a &self.

robin-near commented on chain/client/src/sync/epoch.rs:
```diff
@@ -514,6 +508,148 @@ impl EpochSync {
+    ) -> Result<(), Error> {
+        // Verify first_block_header_in_epoch
+        let first_block_header = &current_epoch.first_block_header_in_epoch;
+        if !verify_hash(
```
nit: rename it verify_block_merkle_proof, or explicitly mention the module name here, for readability? Without the context of "merkle", `verify_hash` doesn't mean anything.

robin-near commented on chain/client/src/sync/epoch.rs:
```diff
@@ -514,6 +508,148 @@ impl EpochSync {
+        epoch_manager: &dyn EpochManagerAdapter,
+    ) -> Result<(), Error> {
+        // Verify epoch.block_producers
+        let bp_hash = Chain::compute_bp_hash_from_validator_stakes(
```
Nice refactoring for this function. I was worried earlier about the protocol version check in here, but this is good.

robin-near commented on chain/client/src/sync/epoch.rs:
```diff
@@ -514,6 +508,148 @@ impl EpochSync {
     }
+
+    // Verify EpochSyncProof
+    fn verify_proof(
```
Could we add some tests (can add some to the testloop test) that messing with the proof will result in a failing verification? We could, e.g. check that tweaking every byte of a valid serialized proof will cause a failure one way or another. It doesn't cover all possible attacks but at least makes sure that we aren't forgetting any fields.

staffik replied:
Created follow-up issue for tests: https://github.com/near/nearcore/issues/12260.
Might need fixing other issues first.

robin-near commented:
Great implementation!

staffik commented on chain/chain/src/store/merkle_proof.rs:
```diff
@@ -0,0 +1,221 @@
+
+/// Implement block merkle proof retrieval.
+
+pub trait MerkleProofAccess {
```
Exactly, added a comment.

staffik commented:
Addressed PR comments here: https://github.com/near/nearcore/pull/12020/commits/76bdb69f7325632087096aa8df5e6ce35aac7dd0
Needed to add two more fields to the proof.
Also, there are 2 things marked as todo and this PR does not includes tests.
